text/0000-osstring-string-interface.md

```diff
@@ -0,0 +1,457 @@
+platform-specific code.
+
+The method for deciding which portions of an `OsStr` correspond to
+Unicode code points tries to be as inclusive as possible, treating a
+section as Unicode if there is any possible interpretation of it in
+the platform's standard Unicode encoding.
+
+* In Windows, OS strings are sequences of ill-formed UTF-16 code
+  units.  (Rust's internal representation is a WTF-8 encoded string,
+  but, aside from determining what operations can be performed
+  efficiently, this is not exposed in the interface.)  Unpaired
+  surrogates are identified as non-Unicode, and everything else is
+  treated as valid UTF-16.
+
+* In Unix, OS strings are arbitrary byte sequences, which are often
+  interpreted as UTF-8.  A byte is treated as being part of a Unicode
+  section if there is any substring containing that byte that is a
+  valid UTF-8 encoded character.  The self-synchronization property of
+  UTF-8 guarantees that there can be at most one such substring for a
+  given byte.  These code points are treated as Unicode characters,
+  and all other bytes are treated as non-Unicode.  Note that this
+  means that any byte with value less than 128 will be interpreted as
+  Unicode.
+
+## `OsString`
+
+`OsString` will get the following new method:
+```rust
+/// Converts an `OsString` into a `String`, avoiding a copy if possible.
+///
+/// Any non-Unicode sequences are replaced with U+FFFD REPLACEMENT CHARACTER.
+fn into_string_lossy(self) -> String;
+
+```
+
+This is analogous to the existing `OsStr::to_string_lossy` method, but
+transfers ownership.  This operation can be done without a copy if the
+`OsString` contains Unicode data or if the platform is Windows.
+
+## `OsStr`
+
+OsStr will get the following new methods (with supporting code
+and explanations interspersed):
+```rust
+/// Returns an iterator over the Unicode and non-Unicode sections
+/// of the string.  Sections will always be nonempty and Unicode
+/// and non-Unicode sections will always alternate.
+///
+/// # Example
+///
+/// ```
+/// use std::ffi::{OsStr, OsStrSection};
+/// let string = OsStr::new("Hello!");
+/// match string.split_unicode().next().unwrap() {
+///     OsStrSection::Unicode(s) => assert_eq!(s, "Hello!"),
+///     OsStrSection::NonUnicode(s) => panic!("Got non-Unicode: {:?}", s),
+/// }
+/// ```
+fn split_unicode<'a>(&'a self) -> SplitUnicode<'a>;
+
+struct SplitUnicode<'a> { ... }
+impl<'a> Clone for SplitUnicode<'a> { ... }
+impl<'a> Iterator for SplitUnicode<'a> {
+    type Item = OsStrSection<'a>;
+    ...
+}
+impl<'a> DoubleEndedIterator for SplitUnicode<'a> { ... }
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum OsStrSection<'a> {
+    Unicode(&'a str),
+    NonUnicode(&'a OsStr),
+}
+
+```
+
+This provides access to the Unicode and non-Unicode sections of the
+string, as defined above.
+
+
+```rust
+/// Returns true if `needle` is a substring of `self`.
+fn contains_os<S: AsRef<OsStr>>(&self, needle: S) -> bool;
```

Comment:
I think it may be a bit onerous to remember that `contains` and `contains_os` are both methods on an `OsStr`. I think we'd be in a better place (e.g. especially in mirroring `str`) if we could avoid extra methods like this. It may involve perhaps a new `Pattern` trait down below, but we may also be able to substitute `AsRef<OsStr>` for `Pattern` because `str` is in theory far more ubiquitous than `OsStr`